text

mohammadtokko opened the pull request:
feat(user_bundle): handle OTP expiration and improve payment handling

- Add OTP expiration check against `user_order.otp_expired_at` (Supabase) to prevent late verifications
- Use Decimal + ROUND_HALF_UP for card payment currency conversions and cents rounding
- Add background task to update order after promo validation to avoid race conditions
- Fix DCB OTP flow (generate/store/send OTP) and resend logic
- Improve logging and error handling in `app/services/user_service.py`

mohammadtokko committed:
text
feat(user_bundle): handle OTP expiration and improve payment handling

- Add OTP expiration check against `user_order.otp_expired_at` (Supabase) to prevent late verifications
- Use Decimal + ROUND_HALF_UP for card payment currency conversions and cents rounding
- Add background task to update order after promo validation to avoid race conditions
- Fix DCB OTP flow (generate/store/send OTP) and resend logic
- Improve logging and error handling in `app/services/user_service.py`

samah-jamal-monty commented on app/services/user_service.py:
```diff
@@ -334,9 +334,14 @@ async def verify_order_otp(self, user: UserModel, request: VerifyOtpRequestDto)
             raise CustomException(code=404, name=ErrorMessages.OTP_INVALID,
                                   details=ErrorMessages.OTP_INVALID)
 
+        if self.__is_order_otp_expired(user_order.id):
```
``
```suggestion
        if self.__is_order_otp_expired(order_id=user_order.id):
```

mohammadtokko committed:
feat(user_bundle): enforce OTP expiry & harden payment flows

- Validate OTP expiration against `user_order.otp_expired_at` (Supabase) before verification
- Use `Decimal` with `ROUND_HALF_UP` for card currency conversion and cent rounding
- Schedule background task to update order after promo validation to avoid race conditions
- Fix DCB OTP generation, storage, sending and resend logic
- Improve logging and error handling in `UserBundleService`
